src/persist/src/nemesis/generator.rs

```diff
@@ -0,0 +1,81 @@
+// Copyright Materialize, Inc. and contributors. All rights reserved.
+//
+// Use of this software is governed by the Business Source License
+// included in the LICENSE file.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0.
+
+use rand::prelude::SmallRng;
+use rand::{Rng, SeedableRng};
+
+use crate::nemesis::{
+    Input, ReadSnapshotReq, Req, ReqId, SealReq, SnapshotId, TakeSnapshotReq, WriteReq,
+};
+
+pub struct Generator {
+    seed: u64,
+    req_id: u64,
+    snap_id: u64,
+}
+
+// TODO: Make Generator configurable, so individual tests can be written to
+// stress interesting combinations of random traffic.
+impl Generator {
+    pub fn from_seed(seed: u64) -> Self {
+        Generator {
+            seed,
+            req_id: 0,
+            snap_id: 0,
+        }
+    }
+
+    pub fn gen(&mut self) -> Input {
+        let req_id = ReqId(self.req_id);
+        self.req_id += 1;
+        let mut rng = SmallRng::seed_from_u64(self.seed + req_id.0);
+        let req = loop {
+            break match rng.gen_range(0..=3) {
+                0 => {
+                    let stream = rng.gen_range(0..5).to_string();
+                    let key = rng.gen_range('a'..'e').to_string();
+                    let ts = rng.gen_range(0..100);
+                    let diff = if rng.gen_bool(0.9) { 1 } else { -1 };
```

Comment:
Does any part of the persistence code care about the contents of "key" and "diff (e.g. is there compaction, deduplication, etc?) If yes, then a non-uniform distribution over the keys may be useful, as well as diffs outside of -1 and 1 .

Reply:
Yup! Added non-uniform distribution of keys to the big TODO.

Genuine question, do we need diffs other than -1 and 1? AFAIK that's the overwhelming majority of usage and I'd been thinking that'd be enough to stress compaction/consolidation as well as things cancelling out.

Reply:
Might consider adding 2 and -2 because there's some "subtract by one" code and so this might exercise some more edge cases.

Reply:
I would advocate for supporting diff = 0 in write sync, and a mode where we sample diffs between [+1000, -1000] or something like that. I agree that +1 / -1 diffs are expected usage for tables and sources but I think its different when persisting arbitrary operators

Reply:
Okay yeah, this is great! Changed this to a uniform distribution in [-2,2]. Operator persistence is a great point, added a TODO to think about new edge cases to test for that.